Controller/Order/Deny.php

```diff
@@ -0,0 +1,220 @@
+        $quote = $quoteFactory->create()->load($quoteId);
+        if(!is_null($quote)){
+            $message = 'Quote ' . $quoteId . ' is set as denied and sent to Riskified. Additional data saved in database (paymentQuote table). 3D Secure verification level - failed.';
+            //saves 3D Secure Response data in quotePayment table (additional data)
+            $payload['date'] = $currentDate = date('Y-m-d H:i:s', time());
+            $this->updateQuotePaymentDetailsInDb($quote, $payload);
+            //Riskified defined order as fraud - order data is send to Riskified
+            $this->sendDeniedOrderToRiskified($quote);
+            $this->logger->log($message);
+        }else{
+            $message = 'Quote ' . $quoteId . ' cannot be found - cannot send fraud try to Riskified.';
+            $this->logger->log($message);
+        }
+
+        return  $this->resultJsonFactory->create()->setData(['message' => $message]);
+    }
+
+    /**
+     * Saves quote payment details (additional data).
+     * @param $quoteId
+     * @param $paymentDetails
+     * @throws \Exception
+     */
+    protected function updateQuotePaymentDetailsInDb($quote, $paymentDetails)
+    {
+        if(isset($quote)){
+            $this->logger->log('Quote ' . $quote->getEntityId() . ' found - saving Riskified Advise or 3D Secure Response as additional quotePayment data in db.');
+            $quotePayment = $quote->getPayment();
+            $additionalData = $quotePayment->getAdditionalData();
+            //avoid overwriting quotePayment additional data
+            if(is_array($additionalData)){
+                $additionalData['3d_secure'] = $paymentDetails;
+                $additionalData = json_encode($additionalData);
+            }else{
+                $additionalData = ['3d_secure' => $paymentDetails];
+                $additionalData = json_encode($additionalData);
+            }
+            try{
+                $quotePayment->setAdditionalData($additionalData);
+                $quotePayment->save();
+            }catch(RuntimeException $e){
+                $this->logger->log('Cannot save quotePayment additional data ' . $e->getMessage());
+            }
+        }else{
+            $this->logger->log('Quote ' . $quote->getEntityId() . ' not found to save additional quotePayment data in db.');
+        }
+    }
+
+    /**
+     * Sends Denied Quote to Riskified Api
+     */
+    protected function sendDeniedOrderToRiskified($quote)
+    {
+        $orderFactory = $this->orderFactory->create();
+        $order = $orderFactory->loadByAttribute('quote_id', $quote->getEntityId());
+        if(is_numeric($order->getEntityId()) != 1){
+            $order = $quote;
+        }
+        $this->apiOrderLayer->post(
+            $order,
+            Api::ACTION_CHECKOUT_DENIED
+        );
+    }
+
+    /**
+     * Returns unmasked quote id.
+     * @param $cartId
+     * @return int|string
+     */
+    protected function getQuoteId($cartId)
```

Comment:
functions are repeating. Create abstract class with those methods and use it as parent for Deny and Call class.